devito/operator/operator.py

```diff
@@ -517,8 +518,16 @@ def _postprocess_arguments(self, args, **kwargs):
     @cached_property
     def _known_arguments(self):
         """The arguments that can be passed to ``apply`` when running the Operator."""
-        ret = set.union(*[set(i._arg_names) for i in self.input + self.dimensions])
-        return tuple(sorted(ret))
+        ret = set()
+        for i in self.input:
+            ret.update(i._arg_names)
+            try:
+                ret.update(i.grid._arg_names)
+            except AttributeError:
+                pass
+        for d in self.dimensions:
+            ret.update(d._arg_names)
```

Comment:
So:
- `grid` arguments are reprocessed for all inputs?
- `dimensions` are reprocessed one more time afterward?
Seems a bit overkill. By here we should have check the grid is unique so should be something like

```
        for i in self.input:
            ret.update(i._arg_names)
            try:
                ret.update(i.grid._arg_names)
               break
            except AttributeError:
                pass
```

Reply:
I don't follow. This is just the `_known_arguments` property, suitably patched (flawed in current master) so that we catch user errors in supplying arguments to `op.apply(...)`. BTW, it's a `cached_property`, so this runs only once?

And `_arg_names` isn't a recursive call